File: js/src/day2.ts
import {getDataFromFile, getTestFromFile} from "./utils";
import {exec} from "child_process";
import expect from "expect";



const firstStep = (data: string[]): number => {
    let horizontalCounter = 0;
    let depthCounter = 0;
    data.forEach((line, index, array) => {
        const [operator, valueStr] = line.split(" ");
        const value = parseInt(valueStr, 10);
        if (operator === 'forward') {
            horizontalCounter += value;
        } else if (operator === "up") {
            depthCounter -= value;
        } else if (operator === "down") {
            depthCounter += value;
        }
    });
    const result = depthCounter * horizontalCounter;

    return result;
}


const secondStep = (data: string[]): number => {
    let horizontalCounter = 0;
    let depthCounter = 0;
    let aim = 0;
    data.forEach((line, index, array) => {
        const [operator, valueStr] = line.split(" ");
        const value = parseInt(valueStr, 10);
        if (operator === 'forward') {
            horizontalCounter += value;
            depthCounter += aim * value
        } else if (operator === "up") {
            aim -= value;
        } else if (operator === "down") {
            aim += value;
        }
    });
    const result = depthCounter * horizontalCounter;

    return result;
}


// EXEC
const dataTest = getTestFromFile("day2");
expect(firstStep(dataTest)).toBe(150);
expect(secondStep(dataTest)).toBe(900);

const data = getDataFromFile("day2");
console.log(`First step : ${firstStep(data)}`);
console.log(`Second step : ${secondStep(data)}`);

File: kotlin/app/src/main/kotlin/aoc2021/Day2.kt
package aoc2021;


class Day2() {
    fun firstStep(input: List<Pair<String, Int>>): Int {
        var horizontal = 0;
        var depth = 0;

        input.forEach() {
            if (it.first.equals("forward")) {
                horizontal += it.second;
            } else if(it.first.equals("up")) {
                depth -= it.second;
            } else if(it.first.equals("down")) {
                depth += it.second;
            }
        }

        return horizontal * depth;
    }

    fun secondStep(input: List<Pair<String, Int>>): Int {
        var horizontal = 0;
        var depth = 0;
        var aim = 0;

        input.forEach() {
            if (it.first.equals("forward")) {
                horizontal += it.second;
                depth += it.second  * aim;
            } else if(it.first.equals("up")) {
                aim -= it.second;
            } else if(it.first.equals("down")) {
                aim += it.second;
            }
        }

        return horizontal * depth;
    }
}


fun main() {

    val day2 = Day2();

    val testInput = Utils().readTest("day2")
            .map {
                val (key, map) = it.split(" ", limit = 2);
                Pair(key, map.toInt())
            }
    assert(day2.firstStep(testInput) == 150)
    assert(day2.secondStep(testInput) == 900)

    val input = Utils().readInput("day2").map {
        val (key, map) = it.split(" ", limit = 2);
        Pair(key, map.toInt())
    }
    println(day2.firstStep(input))
    println(day2.secondStep(input))
}
